### PR TITLE
Allow for running CI checks locally using `docker`

### PR DESCRIPTION
**Pull request by TicClick**
closes https://github.com/ppy/osu-wiki/issues/7502

bonus points if you test this on MacOS, as I can not (Ubuntu and WSL work fine though)

## changes in tests

- all tests now understand `--target`, which is used to pass a list of changed files.
  - "a list of changed files" = everything changed in the feature branch + untracked + unstaged + staged but not committed. much more comfortable and reusable.
- the list of changed files is obtained only once, at the beginning of the local script run OR GitHub Action execution. they differ slightly for every step (see the script/workflow)
- `inspect_binary_file_sizes.sh` now checks ALL affected files for three reasons:
  - simplicity
  - the overhead is very small
  - `git ls-files` doesn't work with untracked files which we still want to test

## changes in `run-checks.sh`

- the script builds a Docker image (~500mb) on the first run. all dependencies (`node_modules`, Python packages) are baked in.
- run each test in the container, mounting the osu-wiki repository. I tried hard to save `node_modules` in the container from being "overwritten" by the host directory and I think I succeeded

## TODO

- [x] add steps to compress and minimize the image because `build-essentials` and friends stretch it to ~800 mega fucking bytes (didn't check yet but this might be a Python dependency)
- [x] update our docs
- [x] remove the requirement to commit/stage changes. remark wrapper only works for committed files, which makes this change not as useful as I want it to be

## future improvements

- allow for running arbitrary commands/tests inside the container (for example, outdate files with `osu-wiki-tools check-outdated-files --autofix --autocommit` since the mount point is writable)
- add a script for Windows

### CODE REVIEW

**Review by bdach — COMMENTED** [review]
Runs ok on my machine, but a few things in passing.

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -19,9 +19,23 @@ jobs:
...
+            sort -u < <(
+              # Changes that are not committed (staged + unstaged + untracked), but without deleted files
+              git status --short -v -v --no-renames --porcelain | awk '$1 != "D" { print $2 }'
+              # Changes committed so far (may overlap with the above)
+              git diff --no-renames --name-only --diff-filter=d ${{ github.sha }}^
+            )
+          )
+          echo 'INTERESTING_FILES<<EOF' >> $GITHUB_ENV
+          echo "${INTERESTING_FILES}" >> $GITHUB_ENV
+          echo 'EOF' >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (TicClick) [reply .github/workflows/continuous-integration.yml:34]: yeah I was lazy here

File: scripts/ci/inspect_file_sizes.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        *)
+            FIRST_COMMIT_HASH=$1
+            LAST_COMMIT_HASH=$2
+            # git diff --numstat will output -<TAB>-<TAB>$filename for blobs
+            TARGET_FILES=$( git diff --numstat --no-renames --diff-filter=d ${FIRST_COMMIT_HASH}^ ${LAST_COMMIT_HASH} | awk '/^-\t-\t/ { print $3 }' )
+            MODE="default"
+    esac
+
+    for file in ${TARGET_FILES}; do
+        if test "$MODE" = "default"; then
```
[the remark this answers is not in the document]
Comment (TicClick) [reply scripts/ci/inspect_file_sizes.sh:35]: honestly, I am just more accustomed to `test`, so I plug it everywhere (I also assume it's more portable, which may or may not be true, but at the same time doesn't matter anymore)... prolly need to revisit this

File: run-checks.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+  interesting_articles=$( echo "${interesting_files}" | grep -e ^wiki/ -e ^news/ | grep .md$ )
+
+  if ! ( which docker > /dev/null ); then
+    print_error "Missing Docker. Install it from https://docs.docker.com/engine and restart the shell"
+    exit 1
+  fi
+
+  if test -z $( docker images -q osu-wiki ); then
+    print_warning "No Docker image found, building (this is a one-time procedure)..."
```
[the remark this answers is not in the document]
Comment (TicClick) [reply run-checks.sh:71]: this is one thing I did not consider, thanks. will fix (I also added this after seeing how much context (~2G) it attempts to pass to the build daemon, which is also not true anymore after `DOCKER_BUILDKIT=1` usage)

File: scripts/ci/inspect_file_sizes.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdach) [reply scripts/ci/inspect_file_sizes.sh:35]: Is fine, was mostly curious.

File: scripts/ci/run_remark.sh [hunk 4]
```diff
@@ -17,16 +12,43 @@ if test "${GITHUB_ACTIONS:-}"; then
...
+    -t|--target) TARGET_FILES="${@: 2}" ;;
+    *)    
+      if test $# -gt 2; then
+        print_help
+        exit 1
+      fi
+
+      FIRST_COMMIT_HASH="$1"
+      LAST_COMMIT_HASH="${2:-"$FIRST_COMMIT_HASH"}"
+      TARGET_FILES=$( git diff --diff-filter=d --name-only --no-renames "$FIRST_COMMIT_HASH"^ "$LAST_COMMIT_HASH" -- '*.md' )
```
[the remark this answers is not in the document]
Comment (TicClick) [reply scripts/ci/run_remark.sh:38]: I tested it with only a single file, `Guides/Beginner's_Guide/en.md`, having being edited. I should double-check this by modifying several files and see how it flies

File: .github/workflows/continuous-integration.yml [hunk 5]
```diff
@@ -19,9 +19,26 @@ jobs:
...
+            sort -u < <(
+              # Changes that are not committed (staged + unstaged + untracked), but without deleted files
+              git status --short -v -v --no-renames --porcelain | awk '$1 != "D" { print $2 }'
+              # Changes committed so far (may overlap with the above)
+              git diff --no-renames --name-only --diff-filter=d ${{ github.sha }}^
+            )
+          )
+
+          # https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings
+          DELIMITER="X${RANDOM}"
```
Comment (bdach) [inline .github/workflows/continuous-integration.yml:34]: Where's `${RANDOM}` coming from here? I don't see it set anywhere, and I don't see it on the list of github-provided envvars...

Comment (TicClick) [reply .github/workflows/continuous-integration.yml:34]: it's provided by the shell, an excerpt from `man bash` below:

```
       RANDOM Each  time this parameter is referenced, it expands to a random integer between 0 and 32767.  Assigning
              a value to RANDOM initializes (seeds) the sequence of random numbers.  If RANDOM is unset, it loses its
              special properties, even if it is subsequently reset.
```

File: scripts/ci/run_remark.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (TicClick) [reply scripts/ci/run_remark.sh:38]: I found a flaw in [444923d](https://github.com/ppy/osu-wiki/pull/8187/commits/444923d8989e8453a24d73a18da6c4ac7312f201) but it wasn't related to a funny file name. everything else worked *flaw*lessly

**Review by bdach — DISMISSED** [review]
Seems okay to me, for whatever that's worth.

**Comment by peppy** [comment]
Everything looks to work fine on macOS, except for the use of `readarray` which is not available.

Maybe one of the alternatives in https://stackoverflow.com/questions/23842261/alternative-to-readarray-because-it-does-not-work-on-mac-os-x will work?

```shell
❯ ./run-checks.sh
Preparing the Docker image...
sha256:605a5adafb4f2320f11555df1666a77e6decb0ff615e7ac9e029c4b9228720c1
/Users/dean/Projects/osu-wiki/run-checks.sh: line 21: readarray: command not found
* Skip file size test
/Users/dean/Projects/osu-wiki/run-checks.sh: line 21: readarray: command not found
* Skip article style test
/Users/dean/Projects/osu-wiki/run-checks.sh: line 21: readarray: command not found
* Skip YAML style test
/Users/dean/Projects/osu-wiki/run-checks.sh: line 21: readarray: command not found
* Skip link test
/Users/dean/Projects/osu-wiki/run-checks.sh: line 21: readarray: command not found
* Skip article freshness test
```

**Comment by TicClick** [comment]
should just have fixed that (`tr`-based solution works in my mock script) -- check again?

**Review by peppy — APPROVED** [review]

**Pull request merged**